ext/standard/dns.c

```diff
@@ -169,34 +169,52 @@ PHP_FUNCTION(gethostbyaddr)
 static zend_string *php_gethostbyaddr(char *ip)
 {
 #if HAVE_IPV6 && HAVE_INET_PTON
-	struct in6_addr addr6;
-#endif
-	struct in_addr addr;
-	struct hostent *hp;
+	struct sockaddr_in sa4;
+	struct sockaddr_in6 sa6;
+	char out[NI_MAXHOST], *p;
 
-#if HAVE_IPV6 && HAVE_INET_PTON
-	if (inet_pton(AF_INET6, ip, &addr6)) {
-		hp = gethostbyaddr((char *) &addr6, sizeof(addr6), AF_INET6);
-	} else if (inet_pton(AF_INET, ip, &addr)) {
-		hp = gethostbyaddr((char *) &addr, sizeof(addr), AF_INET);
-	} else {
-		return NULL;
+	if (inet_pton(AF_INET6, ip, &sa6.sin6_addr)) {
+		sa6.sin6_family = AF_INET6;
+
+		if (getnameinfo((struct sockaddr *)&sa6, sizeof(sa6), out, sizeof(out), NULL, 0, 0) < 0) {
+			return zend_string_init(ip, strlen(ip), 0);
+		}
+		p = strchr(out, '%'); /* demangle numeric host with %name suffix */
+		if (p) {
+			*p = 0;
+		}
+		return zend_string_init(out, strlen(out), 0);
+	} else if (inet_pton(AF_INET, ip, &sa4.sin_addr)) {
+		sa4.sin_family = AF_INET;
+
+		if (getnameinfo((struct sockaddr *)&sa4, sizeof(sa4), out, sizeof(out), NULL, 0, 0) < 0) {
+			return zend_string_init(ip, strlen(ip), 0);
```

Comment:
Notice: `getnameinfo` already return numerical ip when not found.
Error (such as DNS failure) are managed the same way to keep consistent behavior with `gethostbyaddr` usage.

Reply:
Changed, as NI_NAMEREQD now return an error when not ound